flutter/flutter
# [tool] Building flavors with native tooling

fixes #155951 

This PR enables building Flutter apps with flavors using native build tooling (Xcode/Gradle).

### Tool part

Implementation is fully based on https://github.com/flutter/flutter/issues/155951#issuecomment-2385017558 (thanks, @andrewkolos !)

### Android part

Android has native support of flavors on the build system level, so no need to any modifications in gradle part.

It's added to assemble here: https://github.com/flutter/flutter/blob/c9923ca609cc772e44b78c9464bdb4891e0fa01d/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy#L1762-L1764

and parsed from current running variant, meaning that running `./gradlew app:assemblePaidDebug` will choose `Paid` flavor.

### iOS part

There is no existing way to parse flavor from Xcode configuration (as far as I know). And FLAVOR is never passed to build if no additional actions performed. My suggestion here is to support 2 ways:

1. Add `FLAVOR` variable to `xcconfig` file (should be done by user). By doing that, `FLAVOR` variable will be passed to build and parsed in `xcode_backend.dart` (or `macOS_assemble.sh`). This way was supported before, obviously.
2. Parse flavor name from configuration variable. I expect configuration to be named as `$CONFIGURATION-$FLAVOR` as suggested in [Creating flavors in iOS and macOS](https://docs.flutter.dev/deployment/flavors#creating-flavors-in-ios-and-macos) documentation.

Having these 2 methods, parsing logic is become similar to one used for parsing `buildMode` here:https://github.com/flutter/flutter/blob/c9923ca609cc772e44b78c9464bdb4891e0fa01d/packages/flutter_tools/bin/xcode_backend.dart#L135

### Testing strategy

I do not know any other method than test this in integration tests:
1. Build with native tooling
2. Run flavor test with `flutter drive --use-application-binary`

Hence tests added only for iOS and Android since ` --use-application-binary` is not supported for another platforms (as per documentation for `flutter drive`).

⚠️ **WARNING** ⚠️: `.ci.yaml` has been modified to enable new tests in presubmit. Then they become green, latest commit in the branch [DO_NOT_MERGE! try run tests](https://github.com/flutter/flutter/pull/157871/commits/fe0c6a34846538608e69ec0c6cac5ea97ad1d15f) should be deleted. In any case, this commit **must be deleted** before merging

## Pre-launch Checklist


If you need help, consider asking for advice on the #hackers-new channel on [Discord].


[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md
